wdl/transforms/draft3/src/main/scala/wdl/draft3/transforms/wdlom2wom/WorkflowDefinitionElementToWomWorkflowDefinition.scala

```diff
@@ -1,19 +1,108 @@
+
+import scalax.collection.Graph
+import scalax.collection.GraphEdge.DiEdge
 
 object WorkflowDefinitionElementToWomWorkflowDefinition {
 
   def convert(a: WorkflowDefinitionElement): ErrorOr[WorkflowDefinition] = {
 
-    val g: ErrorOr[Graph] = Graph.validateAndConstruct(Set.empty)
+    val graphNodeElements: Set[WorkflowGraphNodeElement] = a.graphElements ++ a.inputsSection.toSeq.flatMap(_.inputDeclarations) ++ a.outputsSection.toSeq.flatMap(_.outputs)
 
-    g map { graph =>
+    makeGraph(graphNodeElements, a.name) map { graph =>
       WorkflowDefinition(a.name, graph, Map.empty, Map.empty)
     }
 
   }
+
+  private def makeGraph(graphNodeElements: Set[WorkflowGraphNodeElement], workflowName: String): ErrorOr[WomGraph] = {
+    val unlinkedGraphNodes: Set[UnlinkedGraphNode] = graphNodeElements map UnlinkedGraphNode
+
+    val dependenciesValidation: ErrorOr[List[(UnlinkedGraphNode, UnlinkedGraphNode)]] = generatedByMap(unlinkedGraphNodes) flatMap { gbm =>
+      val nodesConsumed = for {
+        node <- unlinkedGraphNodes
+        consumed <- node.consumedValues
+      } yield (node, consumed)
+
+      nodesConsumed.toList.traverse[ErrorOr, (UnlinkedGraphNode, UnlinkedGraphNode)] { case (node, consumed) =>
+        findGenerator(consumed, gbm) map { g => (g, node) }
+      }
+    }
+
+    def graphNodeCreationFold(currentValidation: ErrorOr[List[GraphNode]], next: UnlinkedGraphNode): ErrorOr[List[GraphNode]] = {
+      currentValidation flatMap { currentList =>
+        val availableValues: Map[String, OutputPort] = (for {
+          node <- currentList
+          port <- node.outputPorts
+        } yield port.name -> port).toMap
+        val nextGraphNodeValidation = UnlinkedGraphNodeElementToGraphNode.convert(GraphNodeMakerInputs(next, availableValues, workflowName))
+        nextGraphNodeValidation map { nextGraphNode => currentList :+ nextGraphNode }
+      }
+    }
+
+    val graphNodesValidation = dependenciesValidation flatMap { getOrdering(unlinkedGraphNodes) } flatMap { ordering: List[UnlinkedGraphNode] =>
+      ordering.foldLeft[ErrorOr[List[GraphNode]]](List.empty[GraphNode].validNel)(graphNodeCreationFold)
+    }
+
+    println(graphNodesValidation)
+
+    graphNodesValidation flatMap { graphNodes => WomGraph.validateAndConstruct(graphNodes.toSet) }
+  }
+
+  private def getOrdering(unlinkedGraphNodes: Set[UnlinkedGraphNode])
+                         (dependencies: List[(UnlinkedGraphNode, UnlinkedGraphNode)]): ErrorOr[List[UnlinkedGraphNode]] = {
+    // Find the topological order in which we must create the graph nodes:
+    val edges = dependencies map { case (from, to) => DiEdge(from, to) }
+
+    Graph.from[UnlinkedGraphNode, DiEdge](unlinkedGraphNodes, edges).topologicalSort match {
+      case Left(cycleNode) => s"This workflow contains a cyclic dependency on ${cycleNode.value}".invalidNel
+      case Right(topologicalOrder) => topologicalOrder.toList.map(_.value).validNel
```

Comment:
I'm getting a compilation error here due to type mismatch.

Reply:
A compile error or an intelliJ squiggle? I get a red highlight but the compiler is happy. Maybe a comment would be useful though

Reply:
It's like a Heisenberg compile error. It only fails to compile if I'm looking at that line.